### ISSUE TITLE
Fail to load wasm filter built by tinygo

### ISSUE THREAD

**Issue by Neil-Boyle** [envoyproxy/envoy#20483]
**Fail to load wasm filter built by tinygo**

*Description*:

I used tinygo to build wasm filter
command :	@tinygo build -o ./build/envoy_filter.wasm -scheduler=none -target=wasi ./cmd/main.go

errors occured when i starting envoy 

ctitical error info:
` Failed to load Wasm module due to a missing import: wasi_snapshot_preview1.fd_prestat_get`

detail info:

```
[2022-03-23 15:40:48.651][25551][error][wasm] [external/envoy/source/extensions/common/wasm/wasm_vm.cc:38] Failed to load Wasm module due to a missing import: wasi_snapshot_preview1.fd_prestat_get
[2022-03-23 15:40:48.651][25551][error][wasm] [external/envoy/source/extensions/common/wasm/wasm_vm.cc:38] Failed to load Wasm module due to a missing import: wasi_snapshot_preview1.fd_prestat_dir_name
[2022-03-23 15:40:48.652][25551][error][wasm] [external/envoy/source/extensions/common/wasm/wasm_vm.cc:38] Failed to load Wasm module due to a missing import: wasi_snapshot_preview1.path_open
[2022-03-23 15:40:48.652][25551][error][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:107] Wasm VM failed Failed to initialize Wasm code
[2022-03-23 15:40:48.652][25567][critical][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:464] Plugin configured to fail closed failed to load
[2022-03-23 15:40:48.652][25569][critical][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:464] Plugin configured to fail closed failed to load
[2022-03-23 15:40:48.652][25551][critical][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:464] Plugin configured to fail closed failed to load
[2022-03-23 15:40:48.652][25551][warning][config] [external/envoy/source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) listener_1: Unable to create Wasm HTTP filter
```

i push this filter path by xds, which in envoy config_dump look like:
```

"http_filters": [
             {
              "name": "envoy.filters.http.wasm",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm",
               "config": {
                "vm_config": {
                 "runtime": "envoy.wasm.runtime.v8",
                 "code": {
                  "local": {
                   "filename": "/tmp/envoy_filter.wasm"
                  }
                 }
                },
                "configuration": {
                 "@type": "type.googleapis.com/google.protobuf.StringValue"
                }
               }
              }
             },
             {
              "name": "envoy.filters.http.router"
             }
            ]

```

tinygo version:
tinygo version 0.22.0 darwin/amd64 (using go version go1.17.2 and LLVM version 13.0.0)

envoy version:
v1.20.1

any help will by welcome!

**Comment by Neil-Boyle** [comment]
I have google it, one suspect of importing _unsupport package_ ,but I travel through all my filter code import; they are

```
	"fmt"
	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm"
	"net/http"
	"strings"
	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm/types"
        "hash/crc32"
	"regexp"
	"github.com/json-iterator/tinygo"
```
used http.Header{}  http.Request{Header: header} 

simple, nothing  special

**Comment by wbpcode** [comment]
cc @PiotrSikora

**Comment by PiotrSikora** [comment]
> envoy version:
> v1.20.1

You need a more recent version of Envoy. This was fixed in v1.21.0 (proxy-wasm/proxy-wasm-cpp-host#201, and envoyproxy/envoy#19074).

**Comment by Neil-Boyle** [comment]
good job! thx!
